plugins/adrv9002.c

```diff
@@ -0,0 +1,1255 @@
+
+	if (!tx->enabled)
+		return;
+
+	combo_box_manual_update(&tx->gain_ctrl);
+	iio_widget_update(&tx->gain);
+	combo_box_manual_update(&tx->ensm);
+	combo_box_manual_update(&tx->port_en);
+	/* generic widgets */
+	iio_update_widgets(tx->w, tx->num_widgets);
+	/* labels */
+	update_label(&tx->rf_bandwidth);
+	update_label(&tx->sampling_rate);
+}
+
+static void rx_sample_rate_update(struct plugin_private *priv)
+{
+	int i;
+
+	for (i = 0; i < priv->n_adcs; i++) {
+		rx_update_device_sampling_freq(priv->adc_name[i],
+					       USE_INTERN_SAMPLING_FREQ);
+	}
+}
+
+static void update_all(struct plugin_private *priv)
+{
+	int i;
+
+	for(i = 0; i < ADRV9002_NUM_CHANNELS; i++) {
+		adrv9002_update_rx_widgets(priv, i);
+		adrv9002_update_tx_widgets(priv, i);
+	}
+	update_label(&priv->temperature);
+	update_dac_manager(priv);
+	rx_sample_rate_update(priv);
+}
+
+static void reload_settings(GtkButton *btn, struct plugin_private *priv)
+{
+	g_source_remove(priv->refresh_timeout);
+	update_all(priv);
+	/* re-arm the timer */
+	priv->refresh_timeout = g_timeout_add(1000, (GSourceFunc)update_display,
+					      priv);
+}
+
+static void adrv9002_check_channel_status(struct plugin_private *priv,
+					  struct iio_channel *channel,
+					  struct adrv9002_common *chan,
+					  const char *gtk_str)
+{
+	int ret;
+	double dummy;
+	/*
+	 * We use this attr to check if the channel is enabled or not since it should only
+	 * return error if the channel is disabled (assuming there's nothing seriously
+	 * wrong with the device).
+	 */
+	ret = iio_channel_attr_read_double(channel, "rf_bandwidth", &dummy);
+	if (ret == -ENODEV) {
+		chan->enabled = false;
+		gtk_widget_hide(GTK_WIDGET(gtk_builder_get_object(priv->builder,
+								  gtk_str)));
+	} else {
+		chan->enabled = true;
+		gtk_widget_show(GTK_WIDGET(gtk_builder_get_object(priv->builder,
+								  gtk_str)));
+	}
+}
+
+void check_enabled_channels(struct plugin_private *priv)
+{
+	int i;
+	struct iio_channel *chann;
+	char chan_str[32];
+	char gtk_str[32];
+
+	for(i = 0; i < ADRV9002_NUM_CHANNELS; i++) {
+		sprintf(chan_str, "voltage%d", i);
+		sprintf(gtk_str, "frame_rx%d", i + 1);
+		/* rx */
+		chann = iio_device_find_channel(priv->adrv9002, chan_str, false);
+		if (chann)
+			adrv9002_check_channel_status(priv, chann, &priv->rx_widgets[i].rx,
+						      gtk_str);
+		/* tx */
+		sprintf(gtk_str, "frame_tx%d", i + 1);
+		chann = iio_device_find_channel(priv->adrv9002, chan_str, true);
+		if (!chann)
+			continue;
+
+		adrv9002_check_channel_status(priv, chann, &priv->tx_widgets[i], gtk_str);
+	}
+}
+
+static void load_profile(GtkFileChooser *chooser, gpointer data)
+{
+	struct plugin_private *priv = data;
+	char *file_name = gtk_file_chooser_get_filename(chooser);
+	FILE *f;
+	char *buf;
+	int ret;
+	ssize_t size;
+
+	f = fopen(file_name, "r");
+	if (!f)
+		goto err;
+
+	fseek(f, 0, SEEK_END);
+	size = ftell(f);
+	rewind(f);
+	buf = malloc(size);
+	if (!buf)
```

Comment:
'f' should get close before jumping to 'err'

Reply:
definitely